crates/oxc_hir/src/hir_util.rs

```diff
@@ -75,3 +77,54 @@ impl<'a, 'b> IsLiteralValue<'a, 'b> for PropertyKey<'a> {
         }
     }
 }
+
+/// port from [closure-compiler](https://github.com/google/closure-compiler/blob/f3ce5ed8b630428e311fe9aa2e20d36560d975e2/src/com/google/javascript/jscomp/AstAnalyzer.java#L94)
+/// Returns true if the node which may have side effects when executed.
+/// This version default to the "safe" assumptions when the compiler object
+/// is not provided (RegExp have side-effects, etc).
+pub trait MayHaveSideEffects<'a, 'b>
+where
+    Self: CheckForStateChange<'a, 'b>,
+{
+    fn may_have_side_effects(&self) -> bool {
+        self.check_for_state_change(false)
+    }
+}
+
+/// port from [closure-compiler](https://github.com/google/closure-compiler/blob/f3ce5ed8b630428e311fe9aa2e20d36560d975e2/src/com/google/javascript/jscomp/AstAnalyzer.java#L241)
+/// Returns true if some node in n's subtree changes application state. If
+/// `check_for_new_objects` is true, we assume that newly created mutable objects (like object
+/// literals) change state. Otherwise, we assume that they have no side effects.
+pub trait CheckForStateChange<'a, 'b> {
+    fn check_for_state_change(&self, check_for_new_objects: bool) -> bool;
+}
+
+impl<'a, 'b> CheckForStateChange<'a, 'b> for Expression<'a> {
```

Comment:
This function incomplete yet, it only works for this PR, I want to port this Function firstly, but I found it's difficult to verify if it is correct. so, I gave up and want to finish it piece by piece.🤣